homeassistant/helpers/service.py

```diff
@@ -362,8 +362,16 @@ async def async_extract_referenced_entity_ids(
         return selected
 
     for ent_entry in ent_reg.entities.values():
-        if ent_entry.area_id in selector.area_ids or (
-            not ent_entry.area_id and ent_entry.device_id in selected.referenced_devices
+        if (
+            ent_entry.area_id in selector.area_ids
+            or (
+                not ent_entry.area_id
+                and ent_entry.device_id in selected.referenced_devices
+            )
+            or (
+                ent_entry.device_id not in selected.missing_devices
+                and ent_entry.device_id in selector.device_ids
```

Comment:
I don't fully understand this. `selector.device_ids` is already part of `referenced_devices`. So now we say, give me all entities that the user picked and are not in missing devices. But why reference missing devices here? Those are devices that are not exist, and so we shouldn't have an entity referencing those either.

Reply:
Frenck suggested this as a safeguard.

Reply:
Yeah, we had troubles understanding it as well, but it makes sense.

Lets say you make this call:

```yaml
- service: light.turn_on
  target:
    device_id: D
```

Where:

- Device `D` = in area `A`
- Device `D` provides entity `E`
- Entity `E` has an overridden area, it is in area `B`.

Device `D` ends up in `referenced_devices` which is great.

Now the **old logic** that resolves the entities to call:

```py
for ent_entry in ent_reg.entities.values():
  if ent_entry.area_id in selector.area_ids or (
    not ent_entry.area_id and ent_entry.device_id in selected.referenced_devices
  ):
     selected.indirectly_referenced.add(ent_entry.entity_id)
```

Result:

- `ent_entry.area_id in selector.area_ids` -> not true, no area is targeted.
- `not ent_entry.area_id` -> not true, entity does have an area.

Consequence: entity service isn't called, even though the device has been specified.

Reply:
I see. Ok, yeah this makes sense then! 

I think that we can skip the safe guard with `missing_devices` because it _shouldn't_ happen.

Reply:
Do you want me to remove the safe guard? All tests pass fine without it.

Reply:
Yes let's remove the safe guard. It's not a situation we are able to enter.